rfcs/040-tei_adapter/README.md

```diff
@@ -68,43 +68,117 @@ The biggest disadvantage is that we don't have control or visibility if the Webh
+commits within a window can return different results depending on whether the query is made before or after a merge. 
+To not loose changes coming from a merge, we would need to chase the commits coming from that branch when we see a merge commit, 
+which in turn means keeping track of the commits we have already seen and effectively keeping track of the
+history of the repo, which is very fiddly and error prone.
+
+Another approach is to ask the GitHub API for trees. Trees are a representation of the file hierarchy in the repo at a 
+certain commit. The following request gives us all the files in the repository at the latest commit in master:
+```
+https://api.github.com/repos/wellcomecollection/wellcome-collection-tei/git/trees/master?recursive=true
+```
 
+This is a small sample of the response:
+```yaml
+{
+"sha": "d947db3638a7ae2aa069e2b19a7db73dfae7d653",
+"url": "https://api.github.com/repos/wellcomecollection/wellcome-collection-tei/git/trees/d947db3638a7ae2aa069e2b19a7db73dfae7d653",
+"tree": [
+  {
+    "path": ".DS_Store",
+    "mode": "100644",
+    "type": "blob",
+    "sha": "ed307046335248d38c18cb8320d6ca9253e0d35b",
+    "size": 22532,
+    "url": "https://api.github.com/repos/wellcomecollection/wellcome-collection-tei/git/blobs/ed307046335248d38c18cb8320d6ca9253e0d35b"
+  },
+  {
+    "path": ".gitattributes",
+    "mode": "100644",
+    "type": "blob",
+    "sha": "bcb42237cafe6a22e80b9d720dcbf6395198d249",
+    "size": 62,
+    "url": "https://api.github.com/repos/wellcomecollection/wellcome-collection-tei/git/blobs/bcb42237cafe6a22e80b9d720dcbf6395198d249"
+  },
+  {
+    "path": "Arabic",
+    "mode": "040000",
+    "type": "tree",
+    "sha": "187c8c750739ee6a8f9f99ebf7420a46457670d5",
+    "url": "https://api.github.com/repos/wellcomecollection/wellcome-collection-tei/git/trees/187c8c750739ee6a8f9f99ebf7420a46457670d5"
+  },
+  {
+    "path": "Arabic/README.md",
+    "mode": "100644",
+    "type": "blob",
+    "sha": "4bfe74311d86293447f173108190a4b4664d68ea",
+    "size": 1753,
+    "url": "https://api.github.com/repos/wellcomecollection/wellcome-collection-tei/git/blobs/4bfe74311d86293447f173108190a4b4664d68ea"
+  },
+  {
+    "path": "Arabic/WMS_Arabic_1.xml",
+    "mode": "100644",
+    "type": "blob",
+    "sha": "17f54c8c5ebccec643e23b54b3f4277f86204282",
+    "size": 59802,
+    "url": "https://api.github.com/repos/wellcomecollection/wellcome-collection-tei/git/blobs/17f54c8c5ebccec643e23b54b3f4277f86204282"
+  },
+...
+  ]
+}
+```
+Every entry in the `tree` can be either a `tree` or a `blob`. A `blob` is a file in the repo and has a hash associated to it. 
+If the content of a file changes, the hash changes. 
+This means that we can query for the tree in master at regular intervals and store the last version of the tree in S3. 
+At every request, we can compare the old tree with the new one to know what changes have happened.
```

Comment:
I'm not sure what the limit is (although [this post suggests quite a lot](https://github.community/t/github-get-tree-api-limits-and-recursivity/1300/2), but I am not sure how big this repo is expected to grow) - but is it worth thinking about?

Reply:
I should have added in the rfc (and I will, thanks). I've checked and now it looks like the limit is 100k files and 7mb in the response. It's not documented in the github api though, I suppose because they want to reserve themselves the opportunity to change that. Tentatively trying, you can use the same endpoint to retrieve all the files in the catalogue-pipeline repo with no issues in one request, so I think we're covered for tei. The entire manuscript collection is about 10k so even if everything is described in tei tomorrow we should still fit within those limits. Anyhow, if the response is truncated, there is a `truncated` flag that we should defnitely check, and implement a more recursive tree fetching strategy. Even in that case, we know the size of the repo so we can always calculate how many requests we will need and it's probably below 10

Reply:
Great, saw this implemented too.